ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala

```diff
@@ -59,19 +61,40 @@ private[apiserver] final class ApiMeteringReportService(
         .fold[Either[String, Option[Ref.ApplicationId]]](Right(None))(t =>
           Ref.ApplicationId.fromString(t).map(Some.apply)
         )
-    } yield {
-      val reportTime = clock()
-      store.getMeteringReportData(from, to, applicationId).map { reportData =>
-        generator.generate(request, from, to, applicationId, reportData, reportTime)
-      }
-    }) match {
-      case Right(f) => f
-      case Left(error) =>
-        Future.failed(
-          ValidationLogger.logFailure(request, ValidationErrors.invalidArgument(error))
-        )
+    } yield (from, to, applicationId)).left.map(ValidationErrors.invalidArgument)
+  }
+
+  private def generateReport(
+      request: GetMeteringReportRequest,
+      from: Timestamp,
+      to: Option[Timestamp],
+      applicationId: Option[Ref.ApplicationId],
+      reportData: ReportData,
+  ): Either[StatusRuntimeException, GetMeteringReportResponse] = {
+    generator.generate(request, from, to, applicationId, reportData, clock()).left.map { e =>
+      LedgerApiErrors.Admin.InternallyInvalidKey.Reject(e).asGrpcError
     }
   }
+
+  override def getMeteringReport(
+      request: GetMeteringReportRequest
+  ): Future[GetMeteringReportResponse] = {
+    logger.info(s"Received metering report request: $request")
+
+    implicit class WrapEither[T](either: Either[StatusRuntimeException, T]) {
```

Comment:
I would go with just a inner method instead of implicit class for this case

Reply:
This _does_ work:

```    implicit def asFuture[T](either: Either[StatusRuntimeException, T]): Future[T] = either.fold(
      e => Future.failed(ValidationLogger.logFailure(request, e)),
      Future.successful,
    )

    for {
      (from, to, applicationId) <- validateRequest(request)
      reportData <- store.getMeteringReportData(from, to, applicationId)
      report <- generateReport(request, from, to, applicationId, reportData)
    } yield report```

Reply:
But it is also more confusing as it it not obvious if the `for` comprehension is working on Either or Future => reverting to original.

Reply:
I meant an explicit method, not an implicit one :) but It's okay, it was not a big concern, more like a taste